DSharpPlus.Commands/Invocation/CommandEmitUtil.cs

```diff
@@ -0,0 +1,142 @@
+    {
+        DynamicMethod dynamicMethod = new
+        (
+            $"{method.Name}-task-wrapper",
+            typeof(Task),
+            [typeof(object), typeof(object?[])]
+        );
+
+        ILGenerator il = dynamicMethod.GetILGenerator();
+
+        EmitMethodWrapper(il, method);
+
+        Func<object?, object?[], Task> taskWrapper = dynamicMethod.CreateDelegate<Func<object?, object?[], Task>>();
+
+        return async (object? instance, object?[] parameters) => await taskWrapper(instance, parameters);
+    }
+
+    // the heavy lifting of emitting IL actually just works entirely the same for both, since we don't do
+    // anything further with the returned Task/ValueTask, and our wrapper is synchronous.
+    private static void EmitMethodWrapper(ILGenerator il, MethodInfo method)
+    {
+        ParameterInfo[] parameters = method.GetParameters();
+        ConstructorInfo argumentExceptionCtor = typeof(ArgumentException).GetConstructor([typeof(string), typeof(string)])!;
+        Label validation = il.DefineLabel();
+
+        // first, verify validity of the parameter array by length. this makes further issues easier to
+        // diagnose, since we'll be getting the actual error when we expect to.
+        il.Emit(OpCodes.Ldarg_1);
+        il.Emit(OpCodes.Ldlen);
+        il.Emit(OpCodes.Conv_I4);
+        il.Emit(OpCodes.Ldc_I4, parameters.Length);
+        il.Emit(OpCodes.Ceq);
+        il.Emit(OpCodes.Ldc_I4_0);
+        il.Emit(OpCodes.Ceq);
+        il.Emit(OpCodes.Stloc_0);
+        il.Emit(OpCodes.Ldloc_0);
+
+        // branch off if the check was successful
+        il.Emit(OpCodes.Brfalse_S, validation);
+
+        // throw if unsuccessful
+        il.Emit(OpCodes.Nop);
+        il.Emit(OpCodes.Ldstr, "args");
+        il.Emit(OpCodes.Ldstr, "Incorrect number of arguments provided.");
+        il.Emit(OpCodes.Newobj, argumentExceptionCtor);
+        il.Emit(OpCodes.Throw);
+
+        // cast and unbox the arguments
+        il.MarkLabel(validation);
+
+        if (!method.IsStatic)
+        {
+            il.Emit(OpCodes.Ldarg_0);
+
+            // not static implies declared within a type
+            if (method.DeclaringType!.IsValueType)
+            {
+                il.Emit(OpCodes.Unbox_Any, method.DeclaringType!);
+            }
+        }
+
+        for (int i = 0; i < parameters.Length; i++)
+        {
+            il.Emit(OpCodes.Ldarg_1);
+            il.Emit(OpCodes.Ldc_I4, i);
+            il.Emit(OpCodes.Ldelem_Ref);
+
+            // unbox value types, cast reference types
+            if (parameters[i].ParameterType.IsValueType)
+            {
+                il.Emit(OpCodes.Unbox_Any, parameters[i].ParameterType);
+            }
+            else
+            {
+                il.Emit(OpCodes.Castclass, parameters[i].ParameterType);
+            }
+        }
+
+        // actually call the command method
+        il.Emit(OpCodes.Call, method);
+
+        il.Emit(OpCodes.Ret);
```

Comment:
Returns the value of the called method or returns with nothing (`void`)?

Reply:
returns the item on the execution stack, which is the return value of the `call`

Reply:
Wonderful.